docs/functionaldatatypes/zset.md

```diff
@@ -0,0 +1,276 @@
+
+```scala mdoc
+val set: ZSet[Int, Boolean] =
+  ZSet.fromSet(Set(1, 2, 3))
+```
+
+If we have an `Iterable` we can convert it into a `ZSet` using the `fromIterable` constructor. The measure of how many times each item appears will just be how many times it appears in the `Iterable`.
+
+```scala mdoc
+val multiSet: MultiSet[Int] =
+ ZSet.fromIterable(List(1, 1, 2))
+```
+
+Finally, we can just construct an empty `ZSet` using the `empty` operator and add elements to it later.
+
+```scala mdoc
+val empty: ZSet[Nothing, Nothing] =
+  ZSet.empty
+```
+
+## Operators on ZSets
+
+As a generalization of sets, `ZSet` supports many of the same operators that exist on sets.
+
+### Set Membership
+
+We can determine how many times an element appears in the set using the `apply` operator. This requires that the measure type has an `Identity` defined for it.
+
+If the element exists in the set the measure of how many times it appears will be returned. Otherwise the identity element will be returned.
+
+Thus, `shoppingList("apples)` will return `2` and `shoppingList("eggs")` will return `0` while `set(1)` will return `true` and `set(4)` will return `false`.
+
+### Set Union
+
+One simple way of combining sets is the union of two sets.
+
+The union of two sets contains all elements that appear in either set. If an element appears in both sets the measure of how many times it appears in the union will be the maximum of the two measures, as determined by an `Ord` defined on the measure type.
+
+```scala mdoc
+val left: ZSet[Int, Int] =
+  ZSet.fromMap(Map(1 -> 2, 2 -> 3, 3 -> 4))
+
+val right: ZSet[Int, Int] =
+  ZSet.fromMap(Map(2 -> 2, 3 -> 3, 4 -> 4))
+
+val union: ZSet[Int, Int] =
+  left.union(right)
+```
+
+Here the result will be `ZSet(1 -> 2, 2 -> 3, 3 -> 4, 4 -> 4)` because all elements in both sets will be included in the union of the two sets.
+
+## Set Intersection
+
+Set intersection is another basic way of combining sets.
+
+The intersection of two sets only contains the elements that appear in both sets. If an element appears in both sets the measure of how many times it appears in the intersection will be the minimum of the two measures, as determined by an `Ord` defined on the measure type.
+
+```scala mdoc:nest
+val left: ZSet[Int, Int] =
+  ZSet.fromMap(Map(1 -> 2, 2 -> 3, 3 -> 4))
+
+val right: ZSet[Int, Int] =
+  ZSet.fromMap(Map(2 -> 2, 3 -> 3, 4 -> 4))
+
+val intersection: ZSet[Int, Int] =
+  left.intersect(right)
+```
+
+The result will be `ZSet(2 -> 2, 3 -> 3)` because the only commonality of the two sets is that they contain these elements.
+
+## Set Difference
+
+We can also compute the difference of two sets.
+
+The difference of two sets is obtained by subtracting the measure of how many times each element appears in the second set from how many times it appears in the first set. This requires an `Inverse` defined on the `Sum` of the measure type to tell us what it means to "subtract" for this type.
```

Comment:
I like this statement as it definitely encourages users to have "algebraic" reasoning/mindset when using Prelude abstractions.

Reply:
Thanks!